### PR TITLE
Mortgage rounding fix

### PR DESCRIPTION
**Pull request by flowers9**
**Bugfix:**

## The Bug
Due to rounding errors in Mortage::Payment(), taking loans in certain amounts can result in a greater (or lesser) total amount of loans being offered.

## Fix Details
Added a new method, Mortage::PrecisePayment(), that returns the unrounded payment schedule (double versus int64_t).  This is then used by Account::Prequalify() and Mortage::Maximum() to accurately calculate the maximum loan available.  Splitting up the loans has no effect on the overall amount that can be loaned out.

This change will have small effects on offered loan compared to what they were originally, but they'll be more consistent now and changes appear to be very small (about .1% in my test case, although a set of carefully crafted loans might see a larger change).

## Testing Done
Combined with my previous PR (allowing the starting player to pay off the loan and then get it, or smaller ones, back), it was easy to watch the total loan offered remain constant.

### CODE REVIEW

**Review by Hurleveur — COMMENTED** [review]
Multiple mortages always felt weird indeed, hopefully this will resolve it
This looks good code wise I'm just worried about the edge case and consistency

File: source/Mortgage.cpp [hunk 1]
```diff
@@ -180,3 +180,17 @@ int64_t Mortgage::Payment() const
...
+
+
+
+// Check the amount of the next payment due.
+double Mortgage::PrecisePayment() const
+{
+	if(!term)
+		return principal;
+	if(!interest)
+		return static_cast<double>(principal) / term;
```
[the remark this answers is not in the document]
Comment (flowers9) [reply source/Mortgage.cpp:192]: Well, they use the same number of bits (on most modern platforms).  However, the maximum value for double is vastly larger than for int64_t (1.79769e+308 versus 9.223372036854775807e+18).  The difference is that int64_t uses all its bits for precision, whereas double sacrifices some precision for dynamic range (i.e., the ability to be much larger or much smaller).

So, it's possible there could be some precision lost at the low end at seriously high income ranges, but I haven't seen income ranges that I think might trigger that (and, honestly, I don't think anyone would care if they get could loans for an extra .00001% of their current net worth).

Also, this change doesn't affect payments (those still round), just the calculation for offering new loans.

[the remark this answers is not in the document]
Comment (Hurleveur) [reply source/Mortgage.cpp:192]: In that case it does make sense because we'd have less imprecision overall

File: source/Mortgage.cpp [hunk 2]
```diff
@@ -27,17 +27,17 @@ using namespace std;
 
 // Find out how much you can afford to borrow with the given annual revenue
 // and the given credit score (which should be between 200 and 800).
-int64_t Mortgage::Maximum(int64_t annualRevenue, int creditScore, int64_t currentPayments)
+int64_t Mortgage::Maximum(int64_t annualRevenue, int creditScore, double currentPayments)
 {
 	const int term = 365;
```
Comment (RisingLeaf) [inline source/Mortgage.cpp:32]: Make this a const static?

File: source/Mortgage.cpp [hunk 3]
```diff
@@ -27,17 +27,17 @@ using namespace std;
...
-	annualRevenue -= term * currentPayments;
-	if(annualRevenue <= 0)
+	double revenue = annualRevenue - term * currentPayments;
```
Comment (RisingLeaf) [inline source/Mortgage.cpp:33]:
```suggestion
	const double revenue = annualRevenue - term * currentPayments;
```
I don't see it getting changed.

File: source/Mortgage.cpp [hunk 4]
```diff
@@ -180,3 +180,17 @@ int64_t Mortgage::Payment() const
...
+
+// Check the amount of the next payment due.
+double Mortgage::PrecisePayment() const
+{
+	if(!term)
+		return principal;
+	if(!interest)
+		return static_cast<double>(principal) / term;
+
+	double power = pow(1. + interest, term);
```
Comment (RisingLeaf) [inline source/Mortgage.cpp:194]:
```suggestion
	const double power = pow(1. + interest, term);
```
gotta love those const

**Review by RisingLeaf — CHANGES_REQUESTED** [review]
Looks good, some minor things

**Comment by flowers9** [comment]
I'm all for the consts.  Only reason I didn't add them earlier was it seemed against the style of the existing code.

File: source/Mortgage.cpp [hunk 5]
```diff
@@ -27,17 +27,17 @@ using namespace std;
...
 // and the given credit score (which should be between 200 and 800).
-int64_t Mortgage::Maximum(int64_t annualRevenue, int creditScore, int64_t currentPayments)
+int64_t Mortgage::Maximum(int64_t annualRevenue, int creditScore, double currentPayments)
 {
-	const int term = 365;
-	annualRevenue -= term * currentPayments;
-	if(annualRevenue <= 0)
+	const static int term = 365;
+	const double revenue = annualRevenue - term * currentPayments;
+	if(revenue <= 0)
```
Comment (warp-core) [inline source/Mortgage.cpp:34]:
```suggestion
	if(revenue <= 0.)
```

File: source/Mortgage.cpp [hunk 6]
```diff
@@ -27,17 +27,17 @@ using namespace std;
 
 // Find out how much you can afford to borrow with the given annual revenue
 // and the given credit score (which should be between 200 and 800).
-int64_t Mortgage::Maximum(int64_t annualRevenue, int creditScore, int64_t currentPayments)
+int64_t Mortgage::Maximum(int64_t annualRevenue, int creditScore, double currentPayments)
 {
-	const int term = 365;
-	annualRevenue -= term * currentPayments;
-	if(annualRevenue <= 0)
+	const static int term = 365;
```
Comment (warp-core) [inline source/Mortgage.cpp:32]:
```suggestion
	static const int term = 365;
```
Though, marking this as `static` won't make a difference after compiler optimisations.

Comment (flowers9) [reply source/Mortgage.cpp]: True enough.  I don't normally mark method invariants as static unless I need the value to stick around between invocations, but I think someone else suggested it.

Comment (RisingLeaf) [reply source/Mortgage.cpp]: Yea me, but you are right warp.

Anyways now that I think about it, this should be moved to the unnamed namespace anyways as it is a constant that is not getting calculated on the flight. That's what we normally do, right?

Comment (Hurleveur) [reply source/Mortgage.cpp]: Yup

Comment (flowers9) [reply source/Mortgage.cpp]: Changed.

**Review by RisingLeaf — APPROVED** [review]

File: source/Account.cpp [hunk 7]
```diff
@@ -390,11 +390,11 @@ void Account::AddFine(int64_t amount)
 // Check how big a mortgage the player can afford to pay at their current income.
 int64_t Account::Prequalify() const
 {
-	int64_t payments = 0;
+	double payments = 0;
```
Comment (warp-core) [inline source/Account.cpp:393]:
```suggestion
	double payments = 0.;
```

**Pull request merged**